samples/tcp/gopher.js

```diff
@@ -0,0 +1,50 @@
+// This is an ESM module implementing a simple Gopher client. Gopher is an old alternative to HTTP,
+// it is also a simple protocol which makes it ideal for demoing TCP sockets in workerd.
+//
+// This module implements both direct and proxied TCP connections. In order to make use of the
+// proxy you'll need to have an HTTP proxy running on port 1234 (this can be changed in config.capnp).
+//
+// Example usage once you have workerd running:
+//
+// - curl localhost:8080/ # Retrieve the main segment in gopher.floodgap.com.
+// - curl localhost:8080/gstats # Retrieve the "gstats" segment.
+// - curl localhost:8080/gstats?use_proxy=1 # Retrieve the "gstats" segment using the configured proxy.
+export default {
+  async fetch(req, env) {
+    const gopherAddr = "gopher.floodgap.com:70";
+
+    const url = new URL(req.url);
+    const useProxy = url.searchParams.has("use_proxy");
+
+    try {
+      const socket = useProxy ? env.proxy.connect(gopherAddr) : connect(gopherAddr);
+
+      // Write data to the socket. Specifically the "selector" followed by CR+LF.
+      const writer = socket.writable.getWriter()
+      const encoder = new TextEncoder();
+      const encoded = encoder.encode(url.pathname + "\r\n");
+      await writer.write(encoded);
+
+      // The Gopher server should now return a response to us and close the connection.
+      // So start reading from the socket.
+      const reader = socket.readable.getReader();
+      const decoder = new TextDecoder();
+
+      let gopherResponse = "";
+      while (true) {
+        const res = await reader.read();
+        if (res.done) {
+          console.log("Stream done, socket connection has been closed.");
```

Comment:
Given that the writer doesn't end, should the socket connection be closed at this point? I would expect a half-open connection here.

Reply:
I would expect the `Socket` to be auto closed whenever it is disconnected. Though this would mean triggering the `closed` promise which seemed tricky (I have a TODO outstanding for that here: https://github.com/cloudflare/workerd/blob/120ddd76781d2b828e686f176f1faa2a2a339a41/src/workerd/api/tcp.h#L82). Having the user close the socket explicitly would at least avoid needing to worry about that.

I'd have to test this but I would expect the writable stream to throw errors if I attempted to start writing to it after the readable stream signifes "done". So implicitly closing the Socket seems to make sense (though its only real effect in that case would be triggering the "closed" promise).

Reply:
hmm, ok... we should think about this carefully. A socket can exist in a half open state where one side is closed but the other is still open.

Specifically... 

```js
const { readable, writable } = connect(...);

const writer = writable.getWriter();
await writer.write('...');
await writer.close();  // readable side is still open.
```

```js
const { readable, writable } = connect(...);

for await (const chunk of readable) {
  console.log(chunk);
}

// Readable side is done, but writable side should still be open.
const writer = writable.getWriter();
await writer.write('...');
await writer.close();
```

In Node.js' API, we have a configuration option `allowHalfOpen` that controls this behavior and I think we should emulate the same behavior. https://nodejs.org/dist/latest-v19.x/docs/api/net.html#new-netsocketoptions

Reply:
Interesting, I haven't seen TCP half open exposed in an API like this before, nor have I ever needed these semantics. Looking around it seems others have wondered similar about the use cases: https://github.com/nodejs/node/issues/36214. 

In any case, I think you are suggesting we emulate the default `allowHalfOpen`=`false` behaviour. i.e. readable side ending also ends the writable side. This will allow us to extend the API in the future with a similar option if necessary.

Just to be more clear about what I intend to implement:

* If readable side is "done" (or `close` is called on it), the writable side also has `close` called on it implicitly.
* If writable side has `close` called on it, the readable side also has `close` called on it implicitly.
* If either writable or readable side are closed then the `closed` promise completes.

Let me know if that's not what you meant.

Reply:
> If writable side has close called on it, the readable side also has close called on it implicitly.

I think that might be different from what Node.js does: https://nodejs.org/dist/latest-v19.x/docs/api/net.html#socketenddata-encoding-callback

Reply:
> In any case, I think you are suggesting we emulate the default allowHalfOpen=false behaviour.

Just the opposite. I think we should emulate Node.js' default behavior and allow half open by default. Specifically..

1. Assuming that the underlying connection is still, in fact, open, then I should be able to close either of my local readable/writable sides without impacting the other.
2. Once the connection does close then yes, both the readable and writable sides should be closed.

So each of those examples I post above should work.

Reply:
Spoke with James about this offline and we agreed to park this discussion for now. We will come back to this later (at the edgeworker integration stage when it's time to write tests for the API).